src/config.rs

```diff
@@ -0,0 +1,164 @@
+    /// Stores a given setting to the storage
+    fn set_setting(&mut self, key: &str, value: Setting);
+    /// Retrieves all the settings in the storage
+    fn get_all_settings(&self) -> HashMap<String, Setting>;
+}
+
+/// Configuration store is the place where the gosub engine can find all configurable options
+pub struct ConfigStore {
+    /// A hashmap of all settings so we can search o(1) time
+    settings: HashMap<String, Setting>,
+    /// A hashmap of all setting descriptions
+    settings_info: HashMap<String, SettingInfo>,
+    /// Keys of all settings so we can iterate keys easily
+    setting_keys: Vec<String>,
+    /// The storage adapter used for persisting and loading keys
+    storage: Box<dyn StorageAdapter>,
+}
+
+impl ConfigStore {
+    /// Creates a new store with the given storage adapter and preloads the store if needed
+    pub fn new(storage: Box<dyn StorageAdapter>, preload: bool) -> Self {
+        let mut store = ConfigStore {
+            settings: HashMap::new(),
+            settings_info: HashMap::new(),
+            setting_keys: Vec::new(),
+            storage,
+        };
+
+        // Populate the settings from the json file
+        store.populate_settings();
+
+        // preload the settings if requested
+        if preload {
+            let all_settings = store.storage.get_all_settings();
+            for (key, value) in all_settings {
+                store.settings.insert(key, value);
+            }
+        }
+
+        store
+    }
+
+    pub fn has(&self, key: &str) -> bool {
+        self.settings.contains_key(key)
+    }
+
+    /// Returns a list of keys that mathces the given search string (can use *)
+    pub fn find(&self, search: &str) -> Vec<String> {
+        let search = WildMatch::new(search);
+
+        let mut keys = Vec::new();
+        for key in &self.setting_keys {
+            if search.matches(key.as_str()) {
+                let key = key.clone();
+                keys.push(key);
+            }
+        }
+
+        keys
+    }
+
+    pub fn get_info(&self, key: &str) -> Option<SettingInfo> {
+        self.settings_info.get(key).cloned()
+    }
+
+    /// Returns the setting with the given key
+    /// If the setting does not exist, it will try and load it from the storage adapter
+    pub fn get(&mut self, key: &str, default: Option<Setting>) -> Setting {
+        if let Some(setting) = self.settings.get(key) {
+            return setting.clone();
+        }
+
+        // Setting not found, try and load it from the storage adapter
+        if let Some(setting) = self.storage.get_setting(key) {
+            self.settings.insert(key.to_string(), setting.clone());
+            return setting.clone();
+        }
+
+        // Panic if we can't find the setting, and we don't have a default
+        if default.is_none() {
+            panic!("Setting {} not found", key);
+        }
+
+        default.unwrap()
+    }
+
+    pub fn set(&mut self, key: &str, value: Setting) {
```

Comment:
We do not do a check to see if the setting type is actually allowed.

For instance, you can set a bool (`b:true`) on a uint typed setting.

Reply:
this has been fixed as well..  it will check the type you add against the type of the default setting.